PublicMapping/district-builder-2
# Add command for publishing a set of static data

## Overview

The `process-geojson` command generates a set of static data for a region. A new command has been created: `publish-region`, which pushes this data to S3 and inserts a row into the `region-config` table.

### Demo

![db-publish-region](https://user-images.githubusercontent.com/6386/80499372-e5849000-893a-11ea-9a1e-1c7223d0a328.png)

### Notes
 
 * I noticed that there were a couple problems with the Dockerfile that caused `process-geojson` to fail. This has been addressed here.
 * The most difficult part of this task by far was getting database access to work in a good way within the CLI. I tried many ways to get this to work, such as using Nest in standalone mode, but unfortunately it doesn't appear that Nest and oclif can function well together. The implementation I finally got to work here uses `typeorm` directly, in a specific way (calling `createConnection` within the oclif `run` function, and using that to get the repository). I think this actually works pretty well in practice. There remains one minor annoyance: any entity that is being used in this manner needs to have explicit type information in the entity definition. So you'll see that I added `type` attributes to the `@Column` decorators in `RegionConfig`. This is another thing I spent a lot of time trying to get around, but it seems that since oclif is being run in javascript-mode, `reflect-metadata` isn't able to figure the types out automatically, which leads to runtime errors.

## Testing Instructions

- Process another set of geojson to test the Dockerfile changes:
  - `./scripts/manage process-geojson data/input/DE.geojson -o data/output-de -n 12,4,4 -x 12,12,12 -l block,tract,county`
- Publish the data:
  - `./scripts/manage publish-region data/output-de US DE Delaware`
- Ensure the data is in S3
  - `aws s3 ls s3://global-districtbuilder-dev-us-east-1/regions/US/DE/` (then drill into your version directory as well)
- Ensure the row has been inserted into the database
  - Log into the site, go to the create project page, and view the XHR requests. There should be a `region-config` item containing the version you just published.

Closes #67
